Enable non-allocating calls to doKinematics, ComputeFramePose, Spatia…

JoshPetersen-TRI opened the pull request:
…lVelocity and SpatialJacobian

- Change checkCachedKinematicsSettings to take in a char* to prevent allocating a std::string
- Redefine transformSpatialMotion to preallocate a maximum of 6 columns for dynamic matrices
- Move geometricJacobian from transformSpatialMotion to explict calculation of frame changes
- Add a new set of ComputeXSpatialVelocityJacobian methods which take as argument the Jacobian to prevent allocation. This includes adding a new geometricJacobian method and findKinematicPath which use preallocated private member variables

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8803)
<!-- Reviewable:end -->

jwnimmer-tri commented:
@drake-jenkins-bot add to whitelist

@JoshPetersen-TRI Apologies, you've hit Drake issue #3537 where the bot spams you.  My apologies!

hongkai-dai commented:
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[multibody/kinematics_cache.h, line 137 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCHhY2hBxGknj3kYRrM:-LCHhY2hBxGknj3kYRrN:bc0itxt) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/kinematics_cache.h#L137)):*
> ```objc
>                                      bool jdot_times_v_required,
>                                      const char* method_name) const;
> ```

I think google style guide treats  a pointer`T*` as a modifiable object. @jwnimmer-tri , do we have an alternative here? The reason why Josh wants to avoid using `std::string` is to prevent dynamic memory allocation.

One alternative I thought is to create an enum class
```C++
enum class MethodName {
};
```
And then overload `operator<<` for this class `MethodName`. Then we do not need to create a `std::string` object through dynamic memory allocation, unless the kinematics cache fails in the checking. But this idea seems cumbersome.

---

*[multibody/rigid_body_tree.h, line 454 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCGsyWHEH37GWHFwipv:-LCGsyWI-y0nHYr5zm08:b6qix5x) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L454)):*
> ```objc
>   /// @param cache Reference to the KinematicsCache.
>   /// @param B Reference to the RigidBody.
>   /// @param X_BF The pose of frame F in body frame B.
> ```

`J_WF` is not documented here.

The documentation is the same as the function above.
````C++
  drake::Matrix6X<T> CalcFrameSpatialVelocityJacobianInWorldFrame(
      const KinematicsCache<T>& cache, const RigidBody<T>& body,
      const drake::Isometry3<T>& X_BF,
      bool in_terms_of_qdot = false) const;
```
The only difference is to move the return type to be an input argument. Instead of copying the documentation, should we just point out that this function does the same thing as `CalcFrameSpatialVelocityJacobianInWorldFrame`, except changing the return type? @jwnimmer-tri what do you think?

---

*[multibody/rigid_body_tree.h, line 461 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCGv5md3WGSqhFL2vpi:-LCGv5md3WGSqhFL2vpj:b8j8j7j) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L461)):*
> ```objc
>   void CalcFrameSpatialVelocityJacobianInWorldFrame(
>       const KinematicsCache<T>& cache, const RigidBody<T>& body,
>       const drake::Isometry3<T>& X_BF, drake::Matrix6X<T>* J_WF,
> ```

Following google style guide, the output `J_WF` should be put after `in_terms_of_qdot`. I would suggest to not to give a default value to `in_terms_of_qdot`, and switch the order of `J_WF` and `in_terms_of_qdot`.

---

*[multibody/rigid_body_tree.h, line 495 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCGukkd-Wmmppj06e_T:-LCGukkd-Wmmppj06e_U:b-fmn6hf) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L495)):*
> ```objc
>   /// time derivative of the generalized position such that
>   /// `V_WF = J_WF * qdot`. `false` for `J_WF` computed with respect to `v`.
>   /// @retval `J_WF`
> ```

No return type. `J_WF` is an input now.

---

*[multibody/rigid_body_tree.h, line 530 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCGv1LMER5YRVnceGRJ:-LCGv1LMER5YRVnceGRK:b-633odd) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L530)):*
> ```objc
>   /// time derivative of the generalized position such that
>   /// `V_WB = J_WB * qdot`. `false` for `J_WB` computed with respect to `v`.
>   /// @retval `J_WB`
> ```

No return type. `J_WB` is an input.

---

*[multibody/rigid_body_tree.h, line 755 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCHj8yK7OzcEGEJYIQR:-LCHj8yK7OzcEGEJYIQS:b-zfwhff) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L755)):*
> ```objc
>                          KinematicPath* path);
> 
>   void FindAncestorBodies(int body_index, std::vector<int>* ancestor_bodies);
> ```

Why this function is not a `const` function?

---

*[multibody/rigid_body_tree.h, line 940 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCHjK6s2M4jpkHmxI62:-LCHjK6t14nbx3s0IlkM:br6lh9r) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L940)):*
> ```objc
>       const KinematicsCache<Scalar>& cache, int base_body_or_frame_ind,
>       int end_effector_body_or_frame_ind, int expressed_in_body_or_frame_ind,
>       drake::Matrix6X<Scalar>* J, bool in_terms_of_qdot = false,
> ```

Switch the order between `in_terms_of_qdot` and `J`.

---

*[multibody/rigid_body_tree.h, line 1625 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCHjTYN3eXCp4NqH-MP:-LCHjTYN3eXCp4NqH-MQ:bybj7lp) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.h#L1625)):*
> ```objc
>   std::vector<std::shared_ptr<RigidBodyFrame<T>>> frames_;
> 
>   // Preallocated scratch pad variables.
> ```

I think these fields should be marked as `mutable`, so that they can be modified in a const function. And also we should add documentation, on why we need to store these data in RigidBodyTree, which function would use these data?

---

*[multibody/rigid_body_tree.cc, line 474 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRql-5-iiwsr1J--Lz:-LCRql-68haj8UUDLTp2:bq9ofh8) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L474)):*
> ```cpp
>   kinematic_path_.joint_direction_signs.reserve(bodies_.size());
>   kinematic_path_.body_path.reserve(bodies_.size());
>   J_positions_ = drake::Matrix6X<T>(6, num_positions_);
> ```

BTW, you could use `J_positions_.resize(6, num_positions_)` here to avoid the copy constructor.

---

*[multibody/rigid_body_tree.cc, line 2027 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRrMpf23XQV1iUpMeE:-LCRrMpf23XQV1iUpMeF:bv6qwc7) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L2027)):*
> ```cpp
> 
> template <typename T>
> std::vector<int> RigidBodyTree<T>::FindAncestorBodies(
> ```

Call `FindAncestorBodies(body_index, &ancestor_body_list)` inside this function to avoid code duplication.

---

*[multibody/rigid_body_tree.cc, line 2051 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRre56-21aHs2n4ayZ:-LCRre56-21aHs2n4ay_:b-lrink2) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L2051)):*
> ```cpp
> 
> template <typename T>
> KinematicPath RigidBodyTree<T>::findKinematicPath(
> ```

Call `findKinematicPath(start_body_or_frame_idx, end_body_or_frame_idx, &path)` in this function.

---

*[multibody/rigid_body_tree.cc, line 2168 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRs7yA95-X13dqnoi9:-LCRs7yA95-X13dqnoiA:b-i3pcwo) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L2168)):*
> ```cpp
> 
>   // compute path
>   path->joint_path.resize(0);
> ```

Will this function will incur dynamic memory allocation, as `joint_path` is resized to 0?

---

*[multibody/rigid_body_tree.cc, line 2255 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRsylz6yfqFuvLTHo0:-LCRsylz6yfqFuvLTHo1:b-tmuwt9) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L2255)):*
> ```cpp
>     auto T_world_to_frame =
>         relativeTransform(cache, expressed_in_body_or_frame_ind, 0);
>     TwistMatrix<Scalar> J_temp(kTwistSize, cols);
> ```

`J_temp` is constructed as dynamic sized matrix here, so will you still have dynamic memory allocation in the code?

Also, could you add a documentation here, that the code block to compute `J_temp` is similar to `transformSpatialMotion`, but `J` could have more than 6 columns, so we do not call `transformSpatialMotion` directly.

---

*[multibody/rigid_body_tree.cc, line 2270 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCRtOM56EPMzDL-6EBo:-LCRtOM56EPMzDL-6EBp:b-wfxfkw) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L2270)):*
> ```cpp
> template <typename T>
> template <typename Scalar>
> void RigidBodyTree<T>::geometricJacobian(const KinematicsCache<Scalar>& cache,
> ```

The other `geometricJacobian` function should call this `geometricJacobian` function, instead of copying and pasting the code.

---

*[multibody/rigid_body_tree.cc, line 3647 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCGtEPED3iziXJeE_OG:-LCGtEPED3iziXJeE_OH:b7ckfdx) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/rigid_body_tree.cc#L3647)):*
> ```cpp
> 
> template <typename T> drake::Matrix6X<T>
> RigidBodyTree<T>::CalcFrameSpatialVelocityJacobianInWorldFrame(
> ```

Can you call `CalcFrameSpatialVelocityJacobianInWorldFrame(cache, body, X_BF, &J, in_terms_of_qdot)` inside this function? There is a lot of code duplication between the two `CalcFrameSpatialVelocityJacobianInWorldFrame` functions.

I would change the definition of this function to
```C++
{
  Matrix6X<T> J(6, in_terms_of_qdot ? get_num_positions() : get_num_velocities());
  CalcFrameSpatialVelocityJacobianInWorldFrame(cache, body, X_BF, &J, in_terms_of_qdot);
  return J;
}
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
*[multibody/kinematics_cache.h, line 137 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCHhY2hBxGknj3kYRrM:-LCTYicb6wLhN03_ddh-:b-r745hc) ([raw file](https://github.com/robotlocomotion/drake/blob/71df089f03b2c77d7c5255f2ee52559f699850c2/multibody/kinematics_cache.h#L137)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

I think google style guide treats  a pointer`T*` as a modifiable object. @jwnimmer-tri , do we have an alternative here? The reason why Josh wants to avoid using `std::string` is to prevent dynamic memory allocation.

One alternative I thought is to create an enum class
```C++
enum class MethodName {
};
```
And then overload `operator<<` for this class `MethodName`. Then we do not need to create a `std::string` object through dynamic memory allocation, unless the kinematics cache fails in the checking. But this idea seems cumbersome.
</blockquote></details>

GSG does not say that pointers are always mutable.  There is nothing wrong with `const char*` as an input argument.  This is a good and correct fix.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803)*
<!-- Sent from Reviewable.io -->

JoshPetersen-TRI commented:
This is ready for review again when you get a chance Hongkai

---

Review status: 3 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LCnxJcr2i1JSMzqeKIr:b-6j9dk4)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Sorry for the delay
+@jwnimmer-tri for platform review please.

---

Reviewed 1 of 5 files at r2.
Review status: 4 of 8 files reviewed at latest revision, all discussions resolved.

---

*[multibody/rigid_body_tree.h, line 1639 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCo0LbT5J-5cT9mZMNf:-LCo0LbT5J-5cT9mZMNg:b-klm1sd) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L1639)):*
> ```objc
>   // intermediate representation since the Jacobian can not be transformed in
>   // place.
>   mutable drake::Matrix6X<T> J_positions_;
> ```

If `J_positions_` and `J_velocities_` are used in just `CalcFrameSpatialVelocityJacobianInWorldFrame`, then I would prefer the name of the variable to suggest where it is being used. Otherwise we will soon run out of variable name, when we have more functions that need to store the Jacobian and velocity. Same for `J_temp_positions`, `J_temp_velocities_` and `v_or_q_indices_`.

You can also put them in a struct
```C++
struct DataInCalcFrameSpatialVelocityJacobianInWorldFrame {
  Matrix6X<T> J_positions_;
  Matrix6X<T> J_velocities_;
};
```

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LD3FxIb4nIx0oEoZJ98:b2o5rfd)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Thanks for the rework so far, for Hongkai's comments.  I've finished my first review pass now.

> I am happy to debate details and/or break this into smaller PRs but I just wanted to make clear the full set of changes needed at this time.

FYI It seems likely that this PR can reach the point of mergability to master.  If you want to split it up to get some pieces into master earlier then I'm fine with that, but I suspect keeping it all together as-one would be best.

---

Reviewed 3 of 6 files at r1, 2 of 5 files at r2.
Review status: 5 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4-EiSBb1OxKCcHz2T:-LD4-EiSBb1OxKCcHz2U:b-4mwyu) (no related file):*
@hongkai-dai Did you consider whether the `J_positions_` and `J_temp...` and similar could or should be fields of the `KinematicsCache`, instead of the `RigidBodyTree`?  Though with only an admittedly cursory look, to me associating them with a KinematicsCache seems to be more consistent with the existing design.

They have the property that their allocation sizes are determined by nq and nv, like other things in KC, and then we don't have to worry about resizing them during RBT cloning, don't have to worry about T being double while Scalar is AutoDiffXd, etc.

---

*[multibody/rigid_body_tree.h, line 525 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3wcj05XzEpm_EOXvt:-LD3wcj05XzEpm_EOXvu:b-ta0ph5) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L525)):*
> ```objc
>   /// `V_WB = J_WB * v`, where `v` is the generalized velocity. This version
>   /// does not allocate memory and will assert if `J_WB` is incorrectly sized.
>   /// @param[in]cache Reference to the KinematicsCache.
> ```

BTW missing space `[in]cache`.

---

*[multibody/rigid_body_tree.h, line 741 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3wsto9yzctzKKt6HS:-LD3wstp6zvrwgINYX7j:b-t2b8cx) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L741)):*
> ```objc
>    */
>   void FindAncestorBodies(int body_index,
>                           std::vector<int>* ancestor_bodies) const;
> ```

For `rigid_body_tree.h`, we try not to make flag-day changes to the API -- we add new methods and (possibly) deprecate then later remove old ones.  This new method is fine, but please leave the old one also intact (delegating to this one).

---

*[multibody/rigid_body_tree.h, line 751 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3xF3P3CUd3KbIhng8:-LD3xF3P3CUd3KbIhng9:b-tes4o) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L751)):*
> ```objc
>   void findKinematicPath(int start_body_or_frame_idx,
>                          int end_body_or_frame_idx,
>                          KinematicPath* path) const;
> ```

For `rigid_body_tree.h`, we try not to make flag-day changes to the API -- we add new methods and (possibly) deprecate then later remove old ones.  This new method is fine, but please leave the old one also intact (delegating to this one).

Note that the new method should follow GSG so `FindKinematicPath` (cap `F`).

Note then that some of the unit test disturbances can be reverted.

---

*[multibody/rigid_body_tree.h, line 928 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD435T_7UKBmXAy73zM:-LD435T_7UKBmXAy73zN:b-yoyoqp) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L928)):*
> ```objc
>   void transformJacobian(
>       const Eigen::Transform<Scalar, 3, Eigen::Isometry>& transform,
>       bool in_terms_of_qdot, drake::Matrix6X<Scalar>* J) const;
> ```

Public methods need documentation per GSG.

In this particular header, Scalar-templated forward-declared public methods need explicit template instantiations called out in the cc file.

(To fix both -- I suspect this should just be private, instead.)

---

*[multibody/rigid_body_tree.h, line 942 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD46t5--R4e331sZfzv:-LD46t5--R4e331sZfzw:b-g9kfbi) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L942)):*
> ```objc
>       int end_effector_body_or_frame_ind, int expressed_in_body_or_frame_ind,
>       bool in_terms_of_qdot, std::vector<int>* v_indices,
>       drake::Matrix6X<Scalar>* J) const;
> ```

Public methods need documentation per GSG.

(Perhaps this should just be private, instead?)

---

*[multibody/rigid_body_tree.h, line 1630 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4A7rABOSja4PHELwU:-LD4A7rABOSja4PHELwV:b-ya7918) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L1630)):*
> ```objc
>   // Used in geometricJacobian. Preallocated as the size of the path is
>   // dependent on the base body/frame and end effector body/frame.
>   mutable KinematicPath kinematic_path_;
> ```

I would second Honkai's comment below -- local variables that are escalated to class scope for heap reuse should have super clear and non-reused names, both for de-confliction among different methods and also so that uses of them in the methods themselves  stand out starkly.  In that light, this name seems too short.

Here I would suggest `geometric_jacobian_kinematic_path_;` (or use a struct, `geometric_jacobian_temp_.kinematic_path`).

Then, I would also advocate, within the (_one single_) method that uses this field as-if it were a local, to do `KinematicPath& kinematic_path = long_member_name_;` as an abbreviation, and so that the existing code is undisturbed except for that abbreviation -- after the variable has been introduced, the code within the method doesn't need to "know" where its storage it, it can treat it as-if if were a local.

---

*[multibody/rigid_body_tree.cc, line 202 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD42sRkA7FnQmr-eJh0:-LD42sRkA7FnQmr-eJh1:b-dld0is) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L202)):*
> ```cpp
> 
>   clone->v_or_q_indices_.clear();
>   clone->v_or_q_indices_.reserve(std::max(num_positions_, num_velocities_));
> ```

Brittle duplication of this list with the setup actions in `compile()`.

(It seems to me like this could be a new private helper, that both places call?)

---

*[multibody/rigid_body_tree.cc, line 2045 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD41KCL8zaobL1z5PZQ:-LD41KCM15EKLIPF4fPr:b-jz8yvr) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2045)):*
> ```cpp
>   }
> }
> ```

BTW Spurious vertical whitespace.

---

*[multibody/rigid_body_tree.cc, line 2062 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD41kLz99RxM79GyhrW:-LD41kLz99RxM79GyhrX:be4i5po) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2062)):*
> ```cpp
>   int start_body = parseBodyOrFrameID(start_body_or_frame_idx);
> 
>   start_body_ancestors_.resize(0);
> ```

Its unclear why the resize (or really, `clear()`) isn't done within `FindAncestorBodies`, instead?  At first glance, that seems no less performant but substantially less brittle.

---

*[multibody/rigid_body_tree.cc, line 2100 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4Cz7gDLCKMKsfr7G6:-LD4Cz7h-3P4CW6zpWoc:bojmrvv) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2100)):*
> ```cpp
>   path->joint_path.resize(0);
>   path->joint_direction_signs.resize(0);
>   path->body_path.resize(0);
> ```

BTW Prefer `clear()` to be idiomatic.

---

*[multibody/rigid_body_tree.cc, line 2113 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48ns45vxn36Nbhw_6:-LD48ns45vxn36Nbhw_7:betqcf4) ([raw file](https://github.com/robotlocomotion/drake/blob/398a59e7603cea532d7d9ad863d7fb0d1492f776/multibody/rigid_body_tree.cc#L2113)):*
> ```cpp
>   if (v_or_qdot_indices != nullptr) {
>     v_or_qdot_indices->clear();
>     v_or_qdot_indices->reserve(cols);
> ```

BTW I think the `reserve` was harmless when the user has pre-allocated, but nicely helpful if they have not?  This is a public method, so not all callers will be from within this file.

---

*[multibody/rigid_body_tree.cc, line 2143 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD45Alv6vqjt8wT39OF:-LD45Alw0BJ7b3WU_-p4:bbkhmz8) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2143)):*
> ```cpp
>     J_temp.template bottomRows<3>().noalias() +=
>           transform.linear() * J->template bottomRows<3>();
>     *J = J_temp;
> ```

Given that `J` and `J_temp` are both pre-allocated and identically sized, it is unclear to me why we can't just set `J` directly -- why `J_temp` would be needed.

---

*[multibody/rigid_body_tree.cc, line 2150 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD43OsgAoWwBfQ9r8zl:-LD43OsgAoWwBfQ9r8zm:be3hs68) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2150)):*
> ```cpp
> void RigidBodyTree<double>::transformJacobian(
>   const Eigen::Transform<AutoDiffXd, 3, Eigen::Isometry>& transform,
>   bool /*in_terms_of_qdot*/, Matrix6X<AutoDiffXd>* J) const {
> ```

Unclear why we can ignore the requested `in_terms_of_qdot`.

After further reading, I can see that its because we don't bother with pre-allocated `J` in this case, but that should be made more obvious here.

---

*[multibody/rigid_body_tree.cc, line 2165 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD44eEL4q0OYMX10Wvg:-LD44eEL4q0OYMX10Wvh:b-hf3pt) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2165)):*
> ```cpp
>           transform.linear() * J->template bottomRows<3>();
>     *J = J_temp;
> }
> ```

Pasting the same non-trivial method body twice (modulo a small tweak like the temp storage) is a bad enough smell that it needs to at minimum having a comment to explain why doing so is required.

(I suspect after changes from the other discussions, though, this won't need to be duplicated anyway.)

---

*[multibody/rigid_body_tree.cc, line 2174 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD465dqEcFKrqFRSPXC:-LD465dqEcFKrqFRSPXD:bf2yu9h) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2174)):*
> ```cpp
>     bool in_terms_of_qdot, std::vector<int>* v_or_qdot_indices) const {
>   TwistMatrix<Scalar> J(kTwistSize, in_terms_of_qdot ? get_num_positions()
>                             : get_num_velocities());
> ```

If I am reading the code correctly, this return value now has, e.g., `get_num_positions()` columns (i.e., as many columns as there are dofs in the entire tree), whereas previously it only had as many columns as there were dofs in the kinematic path.  That seems like a big change?  Should we truncate the columns of `J` here to match `v_or_qdot_indices.size`?

---

*[multibody/rigid_body_tree.cc, line 2191 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48K5a0-nqHgqD472Q:-LD48K5a0-nqHgqD472R:bp3kdxb) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2191)):*
> ```cpp
>                                          std::vector<int>* v_or_q_indices,
>                                          Matrix6X<Scalar>* J) const {
>   DRAKE_DEMAND(J != nullptr);
> ```

This lacks a guard that `J` is appropriately sized (so that we don't scribble outside of it, in the below loops).  (The guard either could go here or near `J->setZero()` below.)

---

*[multibody/rigid_body_tree.cc, line 2212 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48bSJ9bkWwM6Cj96s:-LD48bSJ9bkWwM6Cj96t:b-cfpgn9) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2212)):*
> ```cpp
> 
>   if (v_or_q_indices != nullptr) {
>     v_or_q_indices->resize(0);
> ```

BTW Prefer `vector::clear()` over `resize(0)` (as-is on master already).

---

*[multibody/rigid_body_tree.cc, line 3557 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4Bk2RBgnIk1K8_kYo:-LD4Bk2RBgnIk1K8_kYp:bvxmfoo) ([raw file](https://github.com/robotlocomotion/drake/blob/398a59e7603cea532d7d9ad863d7fb0d1492f776/multibody/rigid_body_tree.cc#L3557)):*
> ```cpp
> #endif
> #if DRAKE_RBT_SHARD == 1
> ```

BTW Please keep this blank line intact.

---

*[util/drakeGeometryUtil.h, line 190 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3Gk_d-iTECRWL4-ym:-LD3Gk_e82U_i8Ec0vr7:bjzv9ky) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/util/drakeGeometryUtil.h#L190)):*
> ```objc
>                                   ? 6 : Derived::ColsAtCompileTime>
>       type;
> };
> ```

FYI It would be okay to use `decltype(auto)` for the return type below, in order to remove this struct.  (The struct was written before we had C++14 support.)  This way is fine too, though.

---

*[util/drakeGeometryUtil.h, line 196 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3VK5N30ZT8MGgjkKs:-LD3VK5N30ZT8MGgjkKt:bpzkbnf) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/util/drakeGeometryUtil.h#L196)):*
> ```objc
>     const Eigen::Transform<typename DerivedM::Scalar, 3, Eigen::Isometry>& T,
>     const Eigen::MatrixBase<DerivedM>& M) {
>   DRAKE_ASSERT(M.cols() <= 6);
> ```

Since this method is now restricted to `M`'s with at most 6 columns, the API comment should disclaim so.

However, I think that perhaps _all_ call sites now have exactly 6 columns (or at least, are always fixed size), so another possible resolution here would be to change this to a `static_assert` instead of a runtime one.  (The `static_assert` would be self-documenting, and thus not need an API comment change.)

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LD41xiJ-KIMq0GPCSUU:bzez65y)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
Reviewed 3 of 5 files at r2, 5 of 9 files at r3.
Review status: 5 of 6 files reviewed at latest revision, 22 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803)*
<!-- Sent from Reviewable.io -->

JoshPetersen-TRI commented:
Thanks for the review. With your suggestions, the PR has reduced somewhat so I think it is best to leave as one.

---

Review status: 5 of 6 files reviewed at latest revision, 22 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4-EiSBb1OxKCcHz2T:-LDI1jdVBHmG_5ob9FzV:b-dcmx6v) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@hongkai-dai Did you consider whether the `J_positions_` and `J_temp...` and similar could or should be fields of the `KinematicsCache`, instead of the `RigidBodyTree`?  Though with only an admittedly cursory look, to me associating them with a KinematicsCache seems to be more consistent with the existing design.

They have the property that their allocation sizes are determined by nq and nv, like other things in KC, and then we don't have to worry about resizing them during RBT cloning, don't have to worry about T being double while Scalar is AutoDiffXd, etc.
</blockquote></details>

Good point, moved the added preallocated fields to public mutable fields of KinematicsCache

---

*[multibody/rigid_body_tree.h, line 525 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3wcj05XzEpm_EOXvt:-LDCkBNy8FxeYtmRRpMw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L525)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW missing space `[in]cache`.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 741 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3wsto9yzctzKKt6HS:-LDDAkhq13oMk0tUr89E:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L741)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

For `rigid_body_tree.h`, we try not to make flag-day changes to the API -- we add new methods and (possibly) deprecate then later remove old ones.  This new method is fine, but please leave the old one also intact (delegating to this one).
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 751 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3xF3P3CUd3KbIhng8:-LDDCKKU2xfkj4hu56l4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L751)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

For `rigid_body_tree.h`, we try not to make flag-day changes to the API -- we add new methods and (possibly) deprecate then later remove old ones.  This new method is fine, but please leave the old one also intact (delegating to this one).

Note that the new method should follow GSG so `FindKinematicPath` (cap `F`).

Note then that some of the unit test disturbances can be reverted.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 928 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD435T_7UKBmXAy73zM:-LDHsib83Xozo4GRyecG:b4j9rxt) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L928)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Public methods need documentation per GSG.

In this particular header, Scalar-templated forward-declared public methods need explicit template instantiations called out in the cc file.

(To fix both -- I suspect this should just be private, instead.)
</blockquote></details>

Made private.

---

*[multibody/rigid_body_tree.h, line 942 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD46t5--R4e331sZfzv:-LDHsjwW0Duf-KhDCt-u:b-257y03) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L942)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Public methods need documentation per GSG.

(Perhaps this should just be private, instead?)
</blockquote></details>

Made private

---

*[multibody/rigid_body_tree.h, line 1630 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4A7rABOSja4PHELwU:-LDIH4Is1jyHDbulpSz8:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L1630)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I would second Honkai's comment below -- local variables that are escalated to class scope for heap reuse should have super clear and non-reused names, both for de-confliction among different methods and also so that uses of them in the methods themselves  stand out starkly.  In that light, this name seems too short.

Here I would suggest `geometric_jacobian_kinematic_path_;` (or use a struct, `geometric_jacobian_temp_.kinematic_path`).

Then, I would also advocate, within the (_one single_) method that uses this field as-if it were a local, to do `KinematicPath& kinematic_path = long_member_name_;` as an abbreviation, and so that the existing code is undisturbed except for that abbreviation -- after the variable has been introduced, the code within the method doesn't need to "know" where its storage it, it can treat it as-if if were a local.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.h, line 1639 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LCo0LbT5J-5cT9mZMNf:-LDIH31z6t8ZLDyn1ECs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.h#L1639)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

If `J_positions_` and `J_velocities_` are used in just `CalcFrameSpatialVelocityJacobianInWorldFrame`, then I would prefer the name of the variable to suggest where it is being used. Otherwise we will soon run out of variable name, when we have more functions that need to store the Jacobian and velocity. Same for `J_temp_positions`, `J_temp_velocities_` and `v_or_q_indices_`.

You can also put them in a struct
```C++
struct DataInCalcFrameSpatialVelocityJacobianInWorldFrame {
  Matrix6X<T> J_positions_;
  Matrix6X<T> J_velocities_;
};
```
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 202 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD42sRkA7FnQmr-eJh0:-LDHlzF92MQg5OlQ93yt:b7m0pw8) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L202)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Brittle duplication of this list with the setup actions in `compile()`.

(It seems to me like this could be a new private helper, that both places call?)
</blockquote></details>

Removed as these variables have been moved to the KinematicsCache

---

*[multibody/rigid_body_tree.cc, line 2045 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD41KCL8zaobL1z5PZQ:-LDDET262Z8gb7geoloj:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2045)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Spurious vertical whitespace.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 2062 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD41kLz99RxM79GyhrW:-LDDA9kPAD_hBrK61N6c:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2062)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Its unclear why the resize (or really, `clear()`) isn't done within `FindAncestorBodies`, instead?  At first glance, that seems no less performant but substantially less brittle.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 2100 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4Cz7gDLCKMKsfr7G6:-LDDADKD9LPdSRUPF2Bn:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2100)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Prefer `clear()` to be idiomatic.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 2113 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48ns45vxn36Nbhw_6:-LDHzuJN7v4vqksSEoz1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/398a59e7603cea532d7d9ad863d7fb0d1492f776/multibody/rigid_body_tree.cc#L2113)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW I think the `reserve` was harmless when the user has pre-allocated, but nicely helpful if they have not?  This is a public method, so not all callers will be from within this file.
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 2143 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD45Alv6vqjt8wT39OF:-LDJOHrxEYCiturU5xyx:b-o76lzk) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2143)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Given that `J` and `J_temp` are both pre-allocated and identically sized, it is unclear to me why we can't just set `J` directly -- why `J_temp` would be needed.
</blockquote></details>

The intention was to avoid any dynamic sized temporary allocations that Eigen can do under the hood. I tried a for loop over the columns and found this also doesn't allocate so I removed J_temp.

---

*[multibody/rigid_body_tree.cc, line 2150 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD43OsgAoWwBfQ9r8zl:-LDI2aN24sgwY3AGJa8K:b-lfjb86) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2150)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Unclear why we can ignore the requested `in_terms_of_qdot`.

After further reading, I can see that its because we don't bother with pre-allocated `J` in this case, but that should be made more obvious here.
</blockquote></details>

As you figured, duplicate method no longer needed with the move of preallocated variables to KinematicsCache.

---

*[multibody/rigid_body_tree.cc, line 2165 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD44eEL4q0OYMX10Wvg:-LDI25lM6SAI8CWerb73:bpu0h34) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2165)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Pasting the same non-trivial method body twice (modulo a small tweak like the temp storage) is a bad enough smell that it needs to at minimum having a comment to explain why doing so is required.

(I suspect after changes from the other discussions, though, this won't need to be duplicated anyway.)
</blockquote></details>

Duplicate method no longer needed with move of preallocated variables to KinematicsCache.

---

*[multibody/rigid_body_tree.cc, line 2174 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD465dqEcFKrqFRSPXC:-LDIruEG4SvalqvZtHMU:b-1ago43) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2174)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

If I am reading the code correctly, this return value now has, e.g., `get_num_positions()` columns (i.e., as many columns as there are dofs in the entire tree), whereas previously it only had as many columns as there were dofs in the kinematic path.  That seems like a big change?  Should we truncate the columns of `J` here to match `v_or_qdot_indices.size`?
</blockquote></details>

Good point, I thought all users of geometricJacobian transformed to the full positions/velocities size but it does not seem to be the case. Changed geometricJacobian to return the smaller size while GeometricJacobian always returns positions/velocities size to prevent allocations.

---

*[multibody/rigid_body_tree.cc, line 2191 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48K5a0-nqHgqD472Q:-LDI-KV93OvAJXINgTjR:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2191)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This lacks a guard that `J` is appropriately sized (so that we don't scribble outside of it, in the below loops).  (The guard either could go here or near `J->setZero()` below.)
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 2212 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD48bSJ9bkWwM6Cj96s:-LDDCtc45908vz7QuTE1:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/multibody/rigid_body_tree.cc#L2212)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Prefer `vector::clear()` over `resize(0)` (as-is on master already).
</blockquote></details>

Done.

---

*[multibody/rigid_body_tree.cc, line 3557 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD4Bk2RBgnIk1K8_kYo:-LDDD5962cjRiAVIuoe_:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/398a59e7603cea532d7d9ad863d7fb0d1492f776/multibody/rigid_body_tree.cc#L3557)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

BTW Please keep this blank line intact.
</blockquote></details>

Done.

---

*[util/drakeGeometryUtil.h, line 190 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3Gk_d-iTECRWL4-ym:-LDIfsAg7UeRCbQ1Wkxo:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/util/drakeGeometryUtil.h#L190)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

FYI It would be okay to use `decltype(auto)` for the return type below, in order to remove this struct.  (The struct was written before we had C++14 support.)  This way is fine too, though.
</blockquote></details>

Done.

---

*[util/drakeGeometryUtil.h, line 196 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LD3VK5N30ZT8MGgjkKs:-LDIhj1m2fd0vNO1PRQO:b-s2o4c1) ([raw file](https://github.com/robotlocomotion/drake/blob/89385ba6351bb7c084ee22cd4cafdf1ba8decafb/util/drakeGeometryUtil.h#L196)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Since this method is now restricted to `M`'s with at most 6 columns, the API comment should disclaim so.

However, I think that perhaps _all_ call sites now have exactly 6 columns (or at least, are always fixed size), so another possible resolution here would be to change this to a `static_assert` instead of a runtime one.  (The `static_assert` would be self-documenting, and thus not need an API comment change.)
</blockquote></details>

Left as is and added a comment. Currently KinematicsCacheElement motion_subspace_in_body has dynamic columns size (although with a max fixed size of 6) which wouldn't be caught by the static_assert if the max fixed size was changed from 6

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LDOKxWE3bVXUkpJsvHq:b-nmnw4n)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> pending a few open discussions.

(A note to any other Drake platform reviewers who glance over this -- yes, many of the changes here do not meet GSG style, but instead are consistent with the unique prevailing style of these classes in particular.  We could probably err on either more or less compliance on these changes on different axes, but my own judgment was that this style is appropriate for this PR.)

---

Reviewed 6 of 9 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LDOolzN6SCgDWUiW9QV:-LDOolzN6SCgDWUiW9QW:b-fakvrt) (no related file):*
@hongkai-dai could you be sure to take another re-read / update after the changes?  Thanks!  Let us know if / once you have.

---

*[multibody/kinematics_cache.h, line 94 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LDOp9zoAblBhVbOuUZS:-LDOp9zoAblBhVbOuUZT:b-lxlb0h) ([raw file](https://github.com/robotlocomotion/drake/blob/78b2e808973ecea07fbd4a9380ab3128cbf644dc/multibody/kinematics_cache.h#L94)):*
> ```objc
>   // Preallocated variables used in GeometricJacobian. Preallocated as the size
>   // of the path is dependent on the base body/frame and end effector
>   // body/frame.
> ```

BTW Use a doxygen-enabled comment opener (like `///`) for public items (like this struct overview, and ditto below).

---

*[multibody/rigid_body_tree.h, line 746 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LDOnwEk3UbWVXyMZP7-:-LDOnwEk3UbWVXyMZP70:b-iwq6pe) ([raw file](https://github.com/robotlocomotion/drake/blob/78b2e808973ecea07fbd4a9380ab3128cbf644dc/multibody/rigid_body_tree.h#L746)):*
> ```objc
>                           std::vector<int>* ancestor_bodies) const;
> 
>   std::vector<int> FindAncestorBodies(int body_index) const;
> ```

BTW Since this method _had_ nice Doxygen on master, probably we should not let it lose it entirely.

How about something short added here, like "Identical to the above overload, except that is returns the ancestor_bodies instead of using an output argument." or similar?

---

*[multibody/rigid_body_tree.cc, line 2148 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LDOl5jH42V1f729XVWk:-LDOl5jH42V1f729XVWl:b-yxniug) ([raw file](https://github.com/robotlocomotion/drake/blob/78b2e808973ecea07fbd4a9380ab3128cbf644dc/multibody/rigid_body_tree.cc#L2148)):*
> ```cpp
>          -J->col(col).template head<3>().cross(transform.translation());
>     }
> }
> ```

BTW If this method doesn't touch any member fields, for clarity it should at least be marked `static` in the header -- or possibly even simpler to just put it anonymous namespace here and omit the declaration in the header.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LDOoqX1Axgk_R7eT9qe:b-m920dv)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[a discussion](https://reviewable.io/reviews/robotlocomotion/drake/8803#-LDOolzN6SCgDWUiW9QV:-LDOqL4BAw0prS19kbd6:b-9uidiz) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

@hongkai-dai could you be sure to take another re-read / update after the changes?  Thanks!  Let us know if / once you have.
</blockquote></details>

Oh, now I see that you already marked them as reviewed.  Withdrawn.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803)*
<!-- Sent from Reviewable.io -->

hongkai-dai commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8803#-:-LDh4r6q7S6j2Pzc5Kd9:bnfp4nl)*
<!-- Sent from Reviewable.io -->
